QemuPkg/FrontPageButtons/FrontPageButtons.c

```diff
@@ -0,0 +1,253 @@
+    }
+  }
+
+  return TempPtr;
+}
+
+/**
+GetButtonState gets the button state of the Vol+/Vol- buttons from the SMBIOS Table, and stores that
+state in gButtonState.
+
+@return EFI_SUCCESS - String buffer returned to caller
+@return EFI_ERROR   - Error the string
+
+**/
+EFI_STATUS
+GetButtonState (
+  VOID
+  )
+{
+  CHAR8                *BiosString;
+  EFI_SMBIOS_PROTOCOL  *Smbios;
+  EFI_SMBIOS_TYPE      SmbiosType;
+  EFI_SMBIOS_HANDLE    SmbiosHandle;
+  SMBIOS_STRUCTURE     *SmbiosRecord;
+  SMBIOS_TABLE_TYPE3   *SmbiosType3;
+  EFI_STATUS           Status;
+  CHAR8                *StringPtr;
+  UINTN                StrLen;
+
+  Status = EFI_SUCCESS;
+
+  DEBUG ((DEBUG_ERROR, "%a: Entry\n", __FUNCTION__));
+
+  Status = gBS->LocateProtocol (&gEfiSmbiosProtocolGuid, NULL, (VOID *)&Smbios);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Unable to locate SmBiosProtocol.  Code=%r\n", __FUNCTION__, Status));
+    goto Exit;
+  }
+
+  SmbiosType   = SMBIOS_TYPE_SYSTEM_ENCLOSURE;
+  SmbiosHandle = 0xFFFe;
+
+  // -smbios type=3,version=V+
+
+  Status = Smbios->GetNext (Smbios, &SmbiosHandle, &SmbiosType, &SmbiosRecord, NULL);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Unable to get type 3 SMBIOS record.  Code=%r\n", __FUNCTION__, Status));
+    goto Exit;
+  }
+
+  SmbiosType3 = (SMBIOS_TABLE_TYPE3 *)SmbiosRecord;
+  StringPtr   = (CHAR8 *)SmbiosType3 + SmbiosType3->Hdr.Length;
+
+  DEBUG ((DEBUG_INFO, "Type 3 = %p, Size=0x%x, String = %p\n", SmbiosType3, SmbiosType3->Hdr.Length, StringPtr));
+  DUMP_HEX (DEBUG_INFO, 0, SmbiosType3, ((UINTN)StringPtr | 0xFFF) - (UINTN)SmbiosType3, "");
+
+  BiosString = GetBiosString (StringPtr, SmbiosType3->Version);
+
+  if (BiosString != NULL) {
+    StrLen = AsciiStrLen (BiosString);
+
+    if ((StrLen == AsciiStrLen (SMBIOS_VOLUME_UP))  &&
+        (0 == AsciiStrCmp (BiosString, SMBIOS_VOLUME_UP)))
+    {
+      gButtonState = VolUpButton;
+      DEBUG ((DEBUG_INFO, "%a: Vol+ Button Detected\n", __FUNCTION__));
+    }
+
+    if ((StrLen == AsciiStrLen (SMBIOS_VOLUME_DOWN)) &&
+        (0 == AsciiStrCmp (BiosString, SMBIOS_VOLUME_DOWN)))
+    {
+      gButtonState = VolDownButton;
+      DEBUG ((DEBUG_INFO, "%a: Vol- Button Detected\n", __FUNCTION__));
+    }
+  }
+
+  if (gButtonState == NoButtons) {
+    DEBUG ((DEBUG_INFO, "%a: Neither Vol+ nor Vol- detected\n", __FUNCTION__));
+  }
+
+Exit:
+  return Status;
+}
+
+/**
+ Constructor routine to install button services protocol and initialize anything related to buttons
+
+
+@param[in]     - Image Handle of the process loading this module
+@param[in]     - Efi System Table pointer
+
+@retval        - EFI_SUCCESS (always for a constructor)
+ **/
+EFI_STATUS
+EFIAPI
+ButtonsInit (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  MS_BUTTON_SERVICES_PROTOCOL  *Protocol = NULL;
+  EFI_STATUS                   Status    = EFI_SUCCESS;
+
+  DEBUG ((DEBUG_ERROR, "%a \n", __FUNCTION__));

```

Comment:
This should not be error level. Recommend `DEBUG_VERBOSE` (or at least `DEBUG_INFO`).

```suggestion
  DEBUG ((DEBUG_VERBOSE, "%a \n", __FUNCTION__));

```